framework/calico_framework.py

```diff
@@ -0,0 +1,661 @@
+        accepted_offers = []
+        for offer in offers:
+            # Launch tasks onto that offer
+            operation = mesos_pb2.Offer.Operation()
+            operation.type = mesos_pb2.Offer.Operation.LAUNCH
+
+            tasks_for_offer = [task for task in self.tasks if \
+                                task.dependencies_are_met and \
+                                task.slave_id == offer.offer.slave_id.value]
+            for calico_task in tasks_for_offer:
+                global NEXT_AVAILABLE_TASK_ID
+                NEXT_AVAILABLE_TASK_ID += 1
+
+                calico_task.task_id = str(NEXT_AVAILABLE_TASK_ID)
+                calico_task.state = mesos_pb2.TASK_STAGING
+                calico_task.port = offer.port
+
+                _log.info("\t\tLaunching Task %s (%s)", calico_task.task_id, calico_task)
+                _log.debug("\t\tUsing offer %s", offer)
+
+                operation.launch.task_infos.extend([calico_task.as_new_mesos_task()])
+                self.restart_timeout()
+
+            if tasks_for_offer:
+                driver.acceptOffers([offer.offer.id], [operation])
+                accepted_offers.append(offer)
+
+        for offer in accepted_offers:
+            _log.debug("Removing %s", offer)
+            offers.remove(offer)
+
+        # An offer came back which we launched a prior task on.
+        # TODO: Need a slicker way to reserve it outside this function
+        for offer in offers:
+            _log.debug("RESERVING FOR FUTURE: %s, %s", offer, self)
+            if offer.reserved_by == None and \
+                            offer.offer.slave_id.value in [task.slave_id for task in self.tasks]:
+                offer.reserved_by = self
+
+    def timed_out(self, start=False):
+        """
+        Returns true if the test timed out.
+        :param start: Start the timer if it isn't currently running
+        """
+        if self.timeout == None:
+            if start:
+                self.restart_timeout()
+            return False
+        else:
+            return (time.time() - self.timeout) > TEST_TIMEOUT
+
+    def restart_timeout(self):
+        self.timeout = time.time()
+
+
+class Offer(object):
+    """
+    Provides encapsulation around the mesos offer for quick access to mesos
+    settings.
+    """
+    def __init__(self, offer):
+        self.offer = offer
+        self.reserved_by = None
+        self.cpus = 0.0
+        self.mem = 0.0
+        for resource in self.offer.resources:
+            if resource.name == "cpus":
+                self.cpus += resource.scalar.value
+            elif resource.name == "mem":
+                self.mem += resource.scalar.value
+            elif resource.name == "ports":
+                range = resource.ranges.range[:1].pop()
+                self.port = randint(range.begin, range.end)
+
+    def __eq__(self, other):
+        return self.offer.slave_id == other.offer.slave_id
```

Comment:
Each offer is unique by slave. If a new offer comes in, it should replace the old offer from the same slave.